pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,342 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+	"k8s.io/apimachinery/pkg/api/validation"
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/util/shufflesharding"
+	apivalidation "k8s.io/kubernetes/pkg/apis/core/validation"
+	"k8s.io/kubernetes/pkg/apis/flowcontrol"
+)
+
+// ValidateFlowSchemaName validates name for flow-schema.
+var ValidateFlowSchemaName = apimachineryvalidation.NameIsDNSSubdomain
+
+// ValidatePriorityLevelConfigurationName validates name for priority-level-configuration.
+var ValidatePriorityLevelConfigurationName = apimachineryvalidation.NameIsDNSSubdomain
+
+var supportedDistinguisherMethods = sets.NewString(
+	string(flowcontrol.FlowDistinguisherMethodByNamespaceType),
+	string(flowcontrol.FlowDistinguisherMethodByUserType),
+)
+
+var priorityLevelConfigurationQueuingMaxQueues int32 = 10 * 1000 * 1000 // 10^7
+
+var supportedVerbs = sets.NewString(
+	"get",
+	"list",
+	"create",
+	"update",
+	"delete",
+	"deletecollection",
+	"patch",
+	"watch",
+	"proxy",
+)
+
+var supportedSubjectKinds = sets.NewString(
+	string(flowcontrol.SubjectKindServiceAccount),
+	string(flowcontrol.SubjectKindGroup),
+	string(flowcontrol.SubjectKindUser),
+)
+
+var supportedQueuingType = sets.NewString(
+	string(flowcontrol.PriorityLevelQueuingTypeQueueing),
+	string(flowcontrol.PriorityLevelQueuingTypeExempt),
+)
+
+// ValidateFlowSchema validates the content of flow-schema
+func ValidateFlowSchema(fs *flowcontrol.FlowSchema) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&fs.ObjectMeta, false, ValidateFlowSchemaName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateFlowSchemaSpec(&fs.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateFlowSchemaStatus(&fs.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidateFlowSchemaSpec validates the content of flow-schema's spec
+func ValidateFlowSchemaSpec(spec *flowcontrol.FlowSchemaSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if spec.MatchingPrecedence < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("matchingPrecedence"), spec.MatchingPrecedence, "must be positive value"))
+	}
+	if spec.DistinguisherMethod != nil {
+		if !supportedDistinguisherMethods.Has(string(spec.DistinguisherMethod.Type)) {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("distinguisherMethod").Child("type"), spec.DistinguisherMethod, supportedDistinguisherMethods.List()))
+		}
+	}
+	if len(spec.PriorityLevelConfiguration.Name) > 0 {
```

Comment:
Don't we have to check for `spec.PriorityLevelConfiguration == nil` here?

Reply:
the field `PriorityLevelConfiguration`  is a composed struct instead of a pointer. there wont be nil panic here.

Reply:
Oh, right!